src/librustc_metadata/locator.rs

```diff
@@ -892,9 +907,14 @@ fn get_metadata_section_imp(target: &Target,
             }
         }
         CrateFlavor::Rmeta => {
-            let buf = fs::read(filename).map_err(|_|
-                format!("failed to read rmeta metadata: '{}'", filename.display()))?;
-            rustc_erase_owner!(OwningRef::new(buf).map_owner_box())
+            // mmap the file, because only a small fraction of it is read.
+            let file = std::fs::File::open(filename).map_err(|_|
+                format!("failed to open rmeta metadata: '{}'", filename.display()))?;
+            let mmap = unsafe { memmap::Mmap::map(&file) };
```

Comment:
Can we do some simple locking here?

Reply:
My understanding of the problem: we mmap the file's contents into a `Vec` at T1, and that `Vec`'s contents are theoretically set. However, in practice they aren't actually set until they are read, at T2, whereupon the contents will be page-faulted in. But it's possible that the file's contents have changed between T1 and T2. (Even worse, there is actually a different T2 for every separate page in the `Vec`.)

File locking won't help unless we lock the file for the entire period that its contents may be read, which doesn't seem like a good idea.

Also, even if you use the normal file-reading operations, there's still a chance you'll end up with inconsistent data if the file is overwritten in the middle of that operation. But the window of opportunity is potentially much smaller. (I think this is what @alexcrichton is referring to.)

Given that the gains turned out to be minor, I would be ok with abandoning this.

Reply:
> Given that the gains turned out to be minor, I would be ok with abandoning this.

They did? In both time and memory? That's a shame, then.

Reply:
[Here](https://perf.rust-lang.org/compare.html?start=de9666f123e800d5fc34210f23127aa6a5d6e4ef&end=6554f34f7d140e581b0528cc93b9cc30fae1c5bc) is the comparison. `faults` was a big win, but everything else was minimal.

Reply:
That is kind of expected though? It's not like everything depends on lots of large crates, and libstd is not loaded from rmeta at all, so it gets mmap'd by LLVM.

I think it might be more useful to try stage1 `./x.py check` (if at all possible) and/or Servo, as a test subject.

Reply:
Note that the unsafety here is not just about reading inconsistent data -- if the file is truncated while mapped, reads past the new file size (modulo page boundary) will result in sigbus.

Reply:
If we're guaranteed a SIGBUS, it's at least not UB, right?

Reply:
@eddyb As far as interior mutability is concerned that's enough, but I am worried about data races. Even with `UnsafeCell`, the compiler may assume that two subsequent reads of the same location (with no writes to that location happening in between *in the current thread*) return the same result. It may use that in critical situations, like during bounds checks, where it is crucial that we get a consistent value.

AFAIK the only way to opt out of that is to use `AtomicU8`.

Related discussion: https://users.rust-lang.org/t/how-unsafe-is-mmap/19635/

Reply:
I wonder whether we'd get penalized for `AtomicU8` with relaxed reads, heh.

Reply:
For what it's worth, we have a custom crate we use for this in fuchsia called `shared-buffer`, which does a volatile write of a pointer to the buffer in order to prove to LLVM that the memory has escaped, and then only access the memory using ptr reads/writes (`copy_nonoverlapping`), so as to avoid asserting that we have exclusive rights to access the memory. See the code and implementation notes [here](https://fuchsia.googlesource.com/garnet/+/master/public/rust/shared-buffer/src/lib.rs#5) (cc @joshlf).

It doesn't solve the SIGBUS / faulting case, though, which we handle through another mechanism (OS-supported snapshot).

Reply:
I guess I'm not sure of the interactions between volatile and atomic, in general. But I'd rather land something and revisit later (keep an issue open).

Note that we don't need mutation, only reading.

Reply:
Please note that `shared-buffer`'s current approach of using `copy_nonoverlapping` is considered unsound. We intend to move to either volatile or atomic reads/writes, but we've had performance issues, so we haven't done it yet.

Reply:
I *think* `atomic` is want you want here, since this is about memory being shared between different threads (be they in the same process or not) running on the same cache-coherent "domain" of the CPU.

`volatile` is not enough *de jure* because a data race with `volatile` accesses is still a data race and hence UB. However, `volatile` might be good enough in practice.

Reply:
More broadly, what you need is to dissuade the compiler from thinking that the memory you're looking at is vanilla memory, and thus dissuade it from doing any number of optimizations that rely on the assumption that, if it loads the same address twice w/o modifying it in the middle, it will get the same result. Any such optimization would lead to UB if there were, in fact, concurrent modifications. I believe that both a relaxed atomic and a volatile load would do the trick here.

Reply:
(obviously, both volatile and relaxed atomic loads of individual bytes are extremely pessimizing, making it impossible to perform multi-word or even whole-word loads-- you really want a much more specialized tool)

Reply:
That's right. My experience in Rust has been that relaxed atomic loads in a loop perform ~10x worse than non-atomic loads in a loop. I haven't benchmarked volatile loads. If Rust supported volatile memcpy, I suspect it would be a lot better, but yeah, a loop of single-byte volatile loads undoubtedly performs abysmally. You might be able to do better by doing `std::ptr::read_volatile::<T>` where `T` is some multi-byte type like `usize`, but I haven't tried.

Reply:
It would be legal for LLVM to optimize 4 relaxed atomic 1-byte loads into one relaxed atomic 4-byte load. (The same optimization would not be legal for volatile.) *Merging* loads is okay with atomics, just splitting them is not. I am not sure if LLVM does anything like that though.

Reply:
I was hoping it would do something like that, but then my code was 10x slower when I used relaxed atomics :( I didn't spend much time wrestling with it, though, and I'm no optimization expert.

Reply:
Yeah, I expect that there is lots of unused potential around optimizations for atomics. But of course that doesn't help right now, that would take someone to improve LLVM's analyses.

As far as this PR is concerned, the good news is that this is UB only if mutation actually happens. The fact that this is shared is invisible to the program until something else writes to it. That is unlikely to happen. So the end result is that the compiler may fail in arbitrarily unexpected ways *if* someone mutates the mmap'ed file, but absent that nothing strange can happen.

There's no way to ask the kernel to do copy-on-write for an mmap'ed file so that memory, once actually mapped (on the first access), is frozen and not affected by others mutating the file -- is there?

Reply:
Well, the whole point of this PR is that typically only 10--20% of the mapped memory in question is actually read. There are no loops here, so the overhead of atomics are probably not high.